persim/pers_landscape_approx.py

```diff
@@ -0,0 +1,380 @@
+                raise ValueError("start parameter must be passed if values are passed.")
+            if stop is None:
+                raise ValueError("stop parameter must be passed if values are passed.")
+        self.start = start
+        self.stop = stop
+        self.values = values
+        self.num_steps = num_steps
+        if compute:
+            self.compute_landscape()
+
+    def __repr__(self) -> str:
+        return (
+            "The persistence landscape in homological "
+            f"degree {self.hom_deg} on grid from {self.start} to {self.stop}"
+            f" with {self.num_steps} steps"
+        )
+
+    def compute_landscape(self, verbose: bool = False) -> list:
+
+        verboseprint = print if verbose else lambda *a, **k: None
+
+        if self.values.size:
+            verboseprint("values was stored, exiting")
+            return
+
+        verboseprint("values was empty, computing values")
+        # make grid
+        grid_values, step = np.linspace(
+            self.start, self.stop, self.num_steps, retstep=True
+        )
+        # grid_values = list(grid_values)
+        # grid = np.array([[x,y] for x in grid_values for y in grid_values])
+        bd_pairs = self.dgms
+
+        # snap birth-death pairs to grid
+        bd_pairs_grid = ndsnap_regular(bd_pairs, *(grid_values, grid_values))
+
+        # make grid dictionary
+        index = list(range(self.num_steps))
+        dict_grid = dict(zip(grid_values, index))
+
+        # initialze W to a list of 2m + 1 empty lists
+        W = [[] for _ in range(self.num_steps)]
+
+        # for each birth death pair
+        for ind_in_bd_pairs, bd in enumerate(bd_pairs_grid):
+            [b, d] = bd
+            ind_in_Wb = dict_grid[b]  # index in W
+            ind_in_Wd = dict_grid[d]  # index in W
+            mid_pt = (
+                ind_in_Wb + (ind_in_Wd - ind_in_Wb) // 2
+            )  # index half way between, rounded down
+
+            # step through by x value
+            j = 0
+            # j in (b, b+d/2]
+            for _ in range(ind_in_Wb, mid_pt):
+                j += 1
+                # j*step: adding points from a line with slope 1
+                W[ind_in_Wb + j].append(j * step)
+
+            j = 0
+            # j in (b+d/2, d)
+            for _ in range(mid_pt + 1, ind_in_Wd):
+                j += 1
+                W[ind_in_Wd - j].append(j * step)
+
+        # sort each list in W
+        for i in range(len(W)):
+            W[i] = sorted(W[i], reverse=True)
+
+        # calculate k: max length of lists in W
+        K = max([len(_) for _ in W])
+
+        # initialize L to be a zeros matrix of size K x (2m+1)
+        L = np.array([np.zeros(self.num_steps) for _ in range(K)])
+
+        # input Values from W to L
+        for i in range(self.num_steps):
+            for k in range(len(W[i])):
+                L[k][i] = W[i][k]
+
+        # check if L is empty
+        if not L.size:
+            L = np.array(["empty"])
```

Comment:
Is the string "empty" necessary, or can you just do `np.array([])`?

Reply:
This is the case where the grid values entered miss all the birth death pairs, so the persistence landscapes are all zero. However, our algorithm will put the landscape as an empty list. 

We don't want the landscape (values) to be np.array([]) because then the algorithm won't know we already computed landscapes and will just keep computing them every time compute_landscape is called. 

I have pushed a change that puts L as a zeros array instead of np.array(["empty"]).